core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -540,9 +540,11 @@ private[spark] class ExecutorAllocationManager(
 
     private val stageIdToNumTasks = new mutable.HashMap[Int, Int]
     private val stageIdToTaskIndices = new mutable.HashMap[Int, mutable.HashSet[Int]]
-    private val executorIdToTaskIds = new mutable.HashMap[String, mutable.HashSet[Long]]
-    // Number of tasks currently running on the cluster.  Should be 0 when no stages are active.
-    private var numRunningTasks: Int = _
+    private val executorIdToStageAndNumTasks =
+      new mutable.HashMap[String, mutable.HashMap[Int, Int]]
```

Comment:
this is a very complicated data structure. Why do we need to keep track of what stage each executor is running? Is there a simpler way to fix this?